### ISSUE TITLE
activate-hs and uninstall-hs trash cabal executable on Mac OS X

### ISSUE THREAD

**Issue by ygale** [haskell/haskell-platform#147]
With HP 2013.4.0.0 already installed on Mac OS X 10.8.5, I installed 2014.2.0.0. After running `activate-hs` at least once to switch between the two, the `cabal` command just hangs and no longer works.

The installer for 2014.2.0.0 makes `cabal` a symlink to `cabal.wrap`, which is a shell script that ultimately `exec`s `cabal.real`, the actual cabal binary executable. Apparently, older versions of HP structured those differently, and `activate-hs` is a bit confused about that. After the dust settles, the actual cabal binary executable has been overwritten, and all the various `cabal*` files are now either the shell script or a symlink to that same shell script. So an infinite loop is created.

Note that the symlink structure has a few layers of complexity: there are files named `cabal*` in `/usr/bin` and in `/Library/Haskell/ghc-*/bin`, and there is a symlink `/Library/Haskell/current`. These all point to each other either as symlinks or via `exec` in a shell script in various ways. `activate-hs` has to get all of that right for any combination of old and new HP versions.

**Comment by rajasharan** [comment]
I was able to reproduce this issue on a fresh installation of Haskell-Platform 2014.2.0.0 64bit on Mac OS X 10.7.5

```
$> cabal update
...
$> cabal -V
...
$> sudo activate-hs
...
$> cabal -V
(command hangs here forever - takes up 50% of CPU usage and does nothing)
```

**Comment by zhiyuanshi** [comment]
My cabal also hangs forever. Please help, since I can't install any packages. I wonder if there's any recommended quick workaround for this issue?

**Comment by ygale** [comment]
The best way in my opinion is to put a cabal executable somewhere on your path before the Haskell Platform. Anyway, that cabal is very outdated, and the newer versions work fine with an older Platform.

If you don't have any intact cabal executable left on your Mac at all and don't have access to one, then re-install the Haskell Platform. (Which only takes a few minutes, and doesn't cause any problems as long as you always use sandboxes. You do use sandboxes, don't you?) Immediately after that, do `cabal install cabal-install` in a fresh sandbox, and then put the resulting cabal executable on your path before the Haskell Platform.

**Comment by zhiyuanshi** [comment]
@ygale Thanks! Let me have a try.

**Comment by mzero** [comment]
I'm open to better ways of handling cabal on Mac. The main issue is that Cabal's default idea of where things are on a system is baked into the code. It would be great if it has a "system default" config file somewhere... SO - on OS X, I have a wrapper script that sets up the user's .cabal/config file before it runs cabal, and also prints out some other one-time-only info.

The scheme of twisty maze of symlinks doesn't make me happy either... and testing all the old vs. new versions is a hard prospect. .. halp!

**Comment by EarthCitizen** [comment]
This is still an issue even with platform 7.10.3.

**Comment by gbaz** [comment]
I think the bad lines are around here:

https://github.com/haskell/haskell-platform/blob/master/hptool/os-extras/osx/bin/activate-hs#L260

    if [ -x $hpRoot/bin/cabal.wrap ]
    then
        run mv $hpRoot/bin/cabal $hpRoot/bin/cabal.real
        symLink cabal.wrap $hpRoot/bin/cabal
    fi

**Comment by gbaz** [comment]
should be resolved by https://github.com/haskell/haskell-platform/commit/78d92fbe70faae0d7a42f132c8e8ce4a8ca19505

**Comment by gbaz** [comment]
new installers with this fix are now posted for 8.0.1
